src/System.Text.Json/docs/writable_json_dom_spec.md

```diff
@@ -0,0 +1,253 @@
+# Writable JSON Document Object Model (DOM) for `System.Text.Json`
+
+## Introduction
+
+`JsonNode` is a modifiable, dictionary-backed API to complement the readonly `JsonDocument`.
+
+It is the base class for the following concrete types representing all possible kinds of JSON nodes:
+* `JsonString` - representing JSON text value
+* `JsonBoolean` - representing JSON boolean value (`true` or `false`)
+* `JsonNumber` - representing JSON numeric value, can be created from and converted to all possible built-in numeric types
+* `JsonArray` - representing the array of JSON nodes
+* `JsonObject` - representing the set of properties - named JSON nodes
+
+It is a summer internship project being developed by @kasiabulat.
+
+## Goals
+
+The user should be able to:
+* Build up a structured in-memory representation of the JSON payload. 
+* Query the document object model.
+* Modify it. That includes, remove, add, and update. This means we want to build a modifiable JsonDocument analogue that is not just readonly.
+
+## TODOs
+
+* Designing API
+* Implementation of provided methods
+* Tests of provided methods
+* Documentation for public API
+
+## Example scenarios
+### Collection initialization
+
+One of the aims in designing this API was the take an advantage of C# language features and make it easy and natural for delevopers to create instances of `JsonObject`s without calling too many `new` instructions. Below example shows how to initialize JSON object with different types of properties:
+
+```csharp
+var developer = new JsonObject
+{
+    { "name", "Kasia" },
+    { "age", 22 },
+    { "is developer", true },
+    { "null property", (JsonNode) null }
+};
+```
+
+JSON object can be nested within other JSON object or include a JSON array: 
+
+```csharp
+var person = new JsonObject
+{
+    { "name", "John" },
+    { "surname", "Smith" },
+    {
+        "addresses", new JsonObject()
+        {
+            {
+                "office", new JsonObject()
```

Comment:
This is a useful sample.

In addition, it would be interesting to see an equivalent example of JsonObject creation where the values are variables and we have loops/branches/etc. @terrajobst had brought up a good point that in practice, you usually wouldn't creating of JsonObjects like this (with known values/constants).

Similar to something like:
https://github.com/dotnet/core-setup/blob/970b3a5143dfda81556f2b5efc4bfc9c053465eb/src/test/HostActivation.Tests/RuntimeConfig.cs#L49-L79
https://github.com/dotnet/cli/blob/95c6eff6daa1a69f29c42b2d405400ad44bdec91/TestAssets/TestProjects/MultiDependentProject/Program.cs#L11-L23

~So, imagine if the phone numbers were passed-in as a `List<strings>` rather than us hard-coding two values. What would the `JsonArray` node creation look like?~

Edit: The JsonArray sample below showcases it.